### PR TITLE
perf(core) parallelize plugin iterator

### PR DESCRIPTION
**Pull request by kikito**

### CODE REVIEW

**Comment by thibaultcha** [comment]
@kikito We should use the `perf` prefix for this commit message here, as usual.

File: kong/runloop/plugins_iterator.lua [hunk 1]
```diff
@@ -85,129 +85,139 @@ local function load_plugin_configuration(route_id,
...
+  local    route_id = route    and    route.id or nil
+  local  service_id = service  and  service.id or nil
+  local consumer_id = consumer and consumer.id or nil
 
-    local api          = self.api
-    local route        = self.route
-    local service      = self.service
-    local consumer     = ctx.authenticated_consumer
+  local plugin_name = plugin.name
+  local threads = {}
```
Comment (thibaultcha) [inline kong/runloop/plugins_iterator.lua]: Let's preallocate this table with 9 array slots

File: kong/runloop/plugins_iterator.lua [hunk 2]
```diff
@@ -85,129 +85,139 @@ local function load_plugin_configuration(route_id,
...
 
-      if consumer_id then
-        plugin_configuration = load_plugin_configuration(nil, nil, consumer_id, plugin_name, nil)
-        if plugin_configuration then
-          break
-        end
-      end
+    if cfg then
+      return cfg
+    end
```
Comment (thibaultcha) [inline kong/runloop/plugins_iterator.lua]: Seems like this could just be:
```
return cfg
```

(without the surrounding branch)?

Comment (kikito) [reply kong/runloop/plugins_iterator.lua]: No. `cfg` can be `nil` without having an error (if config found for the given parameters on the thread) and we want to keep iterating through the loop in that case.

**Comment by kikito** [comment]
I renamed the commit to `perf` as well

File: kong/runloop/plugins_iterator.lua [hunk 3]
```diff
@@ -85,129 +93,139 @@ local function load_plugin_configuration(route_id,
...
-      if route_id and service_id then
-        plugin_configuration = load_plugin_configuration(route_id, service_id, nil, plugin_name, nil)
-        if plugin_configuration then
-          break
-        end
-      end
+  for i = 1, len do
+    local _, cfg, err = assert(ngx_thread_wait(threads[i]))
+    if err then
+      ngx.ctx.delay_response = false
```
Comment (bungle) [inline kong/runloop/plugins_iterator.lua]: can this just be `ctx.delay_response = false`

File: kong/runloop/plugins_iterator.lua [hunk 4]
```diff
@@ -85,129 +93,139 @@ local function load_plugin_configuration(route_id,
...
-        end
-      end
+  -- perform the most expensive query last, only if needed
+  -- TODO this needs real perf testing to see when it helps.
+  -- Right now it penalizes every single request with global plugins enabled,
+  -- but also uses less CPU. Since we cache negative results inside
+  -- load_plugin_configuration, that could negate the gains
+  local cfg, err = load_plugin_configuration(plugin_name, nil, nil, nil, nil)
+  if err then
+    ngx.ctx.delay_response = false
```
Comment (bungle) [inline kong/runloop/plugins_iterator.lua]: can this just be ctx.delay_response = false

**Comment by thibaultcha** [comment]
Internal testing showed that this was not improving cache churning scenarios after all, and saw a considerable increase of CPU usage. Another attempt without using the `ngx.thread` API on cache hits preserved the performance and resource consumption offered by the current runloop, but made performance under cache churning scenarios worse than that of the current runloop (increased CPU usage and reduced req/s). Closing this as we will explore a few more ideas, and eventually revisit the plugins runloop anyway in the future.

**Pull request closed without merging**